File: util/stage/SCDef.java
package common.util.stage;

import common.battle.StageBasis;
import common.io.InStream;
import common.io.json.JsonClass;
import common.io.json.JsonClass.JCConstructor;
import common.io.json.JsonClass.NoTag;
import common.io.json.JsonField;
import common.io.json.JsonField.GenType;
import common.pack.FixIndexList;
import common.pack.Identifier;
import common.pack.PackData.PackDesc;
import common.system.Copable;
import common.util.Data;
import common.util.unit.AbEnemy;
import common.util.unit.EneRand;
import common.util.unit.Enemy;
import java.util.*;
import java.util.Map.Entry;

@JsonClass
public class SCDef implements Copable<SCDef> {

	@JsonClass(noTag = NoTag.LOAD)
	public static class Line implements Cloneable {
		public Identifier<AbEnemy> enemy;
		public int number, boss, multiple, group;
		public int spawn_0, spawn_1, respawn_0, respawn_1;
		public int castle_0, castle_1, layer_0, layer_1;
		public int mult_atk;

		@JCConstructor
		public Line() {
		}

		public Line(int[] arr) {
			enemy = Identifier.parseInt(arr[E], AbEnemy.class);
			number = arr[N];
			boss = arr[B];
			multiple = arr[M];
			group = arr[G];
			spawn_0 = arr[S0];
			respawn_0 = arr[R0];
			castle_0 = arr[C0];
			layer_0 = arr[L0];
			spawn_1 = arr[S1];
			respawn_1 = arr[R1];
			castle_1 = arr[C1];
			layer_1 = arr[L1];
			mult_atk = arr[M1];
		}

		@Override
		public Line clone() {
			try {
				return (Line) super.clone();
			} catch (CloneNotSupportedException e) {
				e.printStackTrace();
				return null;
			}
		}
	}

	@Deprecated
	public static final int SIZE = 14, E = 0, N = 1, S0 = 2, R0 = 3, R1 = 4, C0 = 5, L0 = 6, L1 = 7, B = 8, M = 9,
			S1 = 10, C1 = 11, G = 12, M1 = 13;

	public static SCDef zread(InStream is) {
		int t = is.nextInt();
		int ver = Data.getVer(is.nextString());
		if (t == 0) {
			if (ver >= 402) {
				int n = is.nextInt();
				int m = is.nextInt();
				SCDef scd = new SCDef(n);
				int[] tmp = new int[SIZE];
				for (int i = 0; i < n; i++) {
					Arrays.fill(tmp, 0);
					for (int j = 0; j < m; j++)
						tmp[j] = is.nextInt();
					if (m < 14)
						tmp[M1] = tmp[M];
					scd.datas[i] = new Line(tmp);
				}
				scd.sdef = is.nextInt();
				n = is.nextInt();
				for (int i = 0; i < n; i++)
					scd.smap.put(Identifier.parseInt(is.nextInt(), AbEnemy.class), is.nextInt());
				n = is.nextInt();
				for (int i = 0; i < n; i++) {
					SCGroup scg = SCGroup.zread(is);
					scd.sub.set(scg.id, scg);
				}
				return scd;
			}
		}
		return null;
	}

	@JsonField
	public Line[] datas;
	@JsonField(gen = GenType.FILL)
	public final FixIndexList<SCGroup> sub = new FixIndexList<>(SCGroup.class);
	@JsonField(generic = { Identifier.class, Integer.class })
	public final TreeMap<Identifier<AbEnemy>, Integer> smap = new TreeMap<>();
	@JsonField
	public int sdef = 0;

	@JCConstructor
	public SCDef() {
	}

	protected SCDef(InStream is, int ver) {
		if (ver >= 305) {
			int n = is.nextByte();
			datas = new Line[n];
			int[] tmp = new int[SIZE];
			for (int i = 0; i < n; i++) {
				Arrays.fill(tmp, 0);
				for (int j = 0; j < 10; j++)
					tmp[j] = is.nextInt();
				tmp[M1] = tmp[M];
				datas[i] = new Line(tmp);
			}
		} else if (ver >= 203) {
			int n = is.nextByte();
			datas = new Line[n];
			int[] tmp = new int[SIZE];
			for (int i = 0; i < n; i++) {
				Arrays.fill(tmp, 0);
				for (int j = 0; j < 10; j++)
					tmp[j] = is.nextInt();
				if (tmp[5] < 100)
					tmp[2] *= -1;
				tmp[M1] = tmp[M];
				datas[i] = new Line(tmp);
			}
		} else
			datas = new Line[0];
	}

	protected SCDef(int s) {
		datas = new Line[s];
	}

	public int allow(StageBasis sb, AbEnemy e) {
		Integer o = smap.get(e.getID());
		o = o == null ? sdef : o;
		if (allow(sb, o, e))
			return o;
		return -1;
	}

	public boolean allow(StageBasis sb, int val, AbEnemy en) {
		if (sb.entityCount(1) >= sb.st.max - ((Enemy)en).de.getWill())
			return false;
		if (val < 0 || val > 1000 || sub.get(val) == null)
			return true;
		SCGroup g = sub.get(val);
		return sb.entityCount(1, val) < g.getMax(sb.est.star);
	}

	public boolean contains(Enemy e) {
		for (Line dat : datas) {
			if(dat.enemy.cls == EneRand.class) {
				EneRand rand = (EneRand) Identifier.get(dat.enemy);

				if(rand != null && rand.contains(e.id, dat.enemy))
					return true;
			} else if(dat.enemy.cls == Enemy.class) {
				if (dat.enemy.equals(e.id))
					return true;
			}
		}
		return false;
	}

	@Override
	public SCDef copy() {
		SCDef ans = new SCDef(datas.length);
		for (int i = 0; i < datas.length; i++)
			ans.datas[i] = datas[i].clone();
		ans.sdef = sdef;
		smap.forEach(ans.smap::put);
		sub.forEach((i, e) -> ans.sub.set(i, e.copy(i)));
		return ans;
	}

	public Set<Enemy> getAllEnemy() {
		Set<Enemy> l = new TreeSet<>();
		for (Line dat : datas)
			l.addAll(Identifier.getOr(dat.enemy, AbEnemy.class).getPossible());
		for (AbEnemy e : getSummon())
			l.addAll(e.getPossible());
		return l;
	}

	public Line[] getSimple() {
		return datas;
	}

	public Line getSimple(int i) {
		return datas[i];
	}

	@SuppressWarnings("unchecked")
	public Entry<Identifier<AbEnemy>, Integer>[] getSMap() {
		return smap.entrySet().toArray(new Entry[0]);
	}

	public Set<AbEnemy> getSummon() {
		Set<AbEnemy> ans = new TreeSet<>();
		Set<AbEnemy> temp = new TreeSet<>();
		Set<Enemy> pre = new TreeSet<>();
		Set<Enemy> post = new TreeSet<>();
		for (Line line : datas) {
			AbEnemy e = Identifier.get(line.enemy);
			if (e != null)
				pre.addAll(e.getPossible());
		}
		while (pre.size() > 0) {
			for (Enemy e : pre)
				temp.addAll(e.de.getSummon());
			ans.addAll(temp);
			post.addAll(pre);
			pre.clear();
			for (AbEnemy e : temp)
				pre.addAll(e.getPossible());
			pre.removeAll(post);
			temp.clear();
		}
		return ans;
	}

	public boolean isSuitable(PackDesc pack) {
		for (Line ints : datas) {
			if (ints.enemy.pack.equals(Identifier.DEF))
				continue;
			boolean b = ints.enemy.pack.equals(pack.id);
			for (String rel : pack.dependency)
				b |= ints.enemy.pack.equals(rel);
			if (!b)
				return false;
		}
		return true;
	}

	public boolean isTrail() {
		for (Line data : datas)
			if (data.castle_0 > 100)
				return true;
		return false;
	}

}

File: util/stage/SCDef.kt
package common.util.stage

import common.battle.StageBasis
import common.io.InStream
import common.io.json.JsonClass
import common.io.json.JsonClass.JCConstructor
import common.io.json.JsonClass.NoTag
import common.io.json.JsonField
import common.io.json.JsonField.GenType
import common.pack.FixIndexList
import common.pack.PackData
import common.pack.PackData.PackDesc
import common.system.Copable
import common.util.Data
import common.util.stage.SCDef
import common.util.stage.SCGroup
import common.util.unit.AbEnemy
import common.util.unit.Enemy
import java.util.*
import java.util.function.BiConsumer

@JsonClass
class SCDef : Copable<SCDef?> {
    @JsonClass(noTag = NoTag.LOAD)
    class Line : Cloneable {
        var enemy: PackData.Identifier<AbEnemy>? = null
        var number = 0
        var boss = 0
        var multiple = 0
        var group = 0
        var spawn_0 = 0
        var spawn_1 = 0
        var respawn_0 = 0
        var respawn_1 = 0
        var castle_0 = 0
        var castle_1 = 0
        var layer_0 = 0
        var layer_1 = 0
        var mult_atk = 0

        @JCConstructor
        constructor() {
        }

        constructor(arr: IntArray) {
            enemy = PackData.Identifier.Companion.parseInt<AbEnemy>(arr[E], Enemy::class.java)
            number = arr[N]
            boss = arr[B]
            multiple = arr[M]
            group = arr[G]
            spawn_0 = arr[S0]
            respawn_0 = arr[R0]
            castle_0 = arr[C0]
            layer_0 = arr[L0]
            spawn_1 = arr[S1]
            respawn_1 = arr[R1]
            castle_1 = arr[C1]
            layer_1 = arr[L1]
            mult_atk = arr[M1]
        }

        public override fun clone(): common.util.stage.SCDef.Line {
            return try {
                super.clone() as common.util.stage.SCDef.Line
            } catch (e: CloneNotSupportedException) {
                e.printStackTrace()
                null
            }
        }
    }

    @JsonField
    var datas: Array<common.util.stage.SCDef.Line?>

    @JsonField(gen = GenType.FILL)
    val sub: FixIndexList<SCGroup> = FixIndexList<SCGroup>(SCGroup::class.java)

    @JsonField(generic = [PackData.Identifier::class, Int::class])
    val smap: MutableMap<PackData.Identifier<AbEnemy>, Int> = TreeMap<PackData.Identifier<AbEnemy>, Int>()

    @JsonField
    var sdef = 0

    protected constructor() {}
    protected constructor(`is`: InStream, ver: Int) {
        if (ver >= 305) {
            val n: Int = `is`.nextByte()
            datas = arrayOfNulls(n)
            val tmp = IntArray(SIZE)
            for (i in 0 until n) {
                Arrays.fill(tmp, 0)
                for (j in 0..9) tmp[j] = `is`.nextInt()
                tmp[M1] = tmp[M]
                datas[i] = common.util.stage.SCDef.Line(tmp)
            }
        } else if (ver >= 203) {
            val n: Int = `is`.nextByte()
            datas = arrayOfNulls(n)
            val tmp = IntArray(SIZE)
            for (i in 0 until n) {
                Arrays.fill(tmp, 0)
                for (j in 0..9) tmp[j] = `is`.nextInt()
                if (tmp[5] < 100) tmp[2] *= -1
                tmp[M1] = tmp[M]
                datas[i] = common.util.stage.SCDef.Line(tmp)
            }
        } else datas = arrayOfNulls(0)
    }

    constructor(s: Int) {
        datas = arrayOfNulls(s)
    }

    fun allow(sb: StageBasis?, e: AbEnemy): Int {
        var o = smap[e.getID()]
        o = o ?: sdef
        return if (allow(sb, o)) o else -1
    }

    fun allow(sb: StageBasis, `val`: Int): Boolean {
        if (sb.entityCount(1) >= sb.st.max) return false
        if (`val` < 0 || `val` > 1000 || sub.get(`val`) == null) return true
        val g: SCGroup = sub.get(`val`)
        return sb.entityCount(1, `val`) < g.getMax(sb.est.star)
    }

    operator fun contains(e: Enemy): Boolean {
        for (dat in datas) if (dat!!.enemy === e.id) return true
        return false
    }

    override fun copy(): SCDef {
        val ans = SCDef(datas.size)
        for (i in datas.indices) ans.datas[i] = datas[i]!!.clone()
        ans.sdef = sdef
        smap.forEach(BiConsumer<PackData.Identifier<AbEnemy>, Int> { e: PackData.Identifier<AbEnemy>, i: Int -> ans.smap.put(e, i) })
        sub.forEach(BiConsumer<Int, SCGroup> { i: Int, e: SCGroup -> ans.sub.set(i, e.copy(i)) })
        return ans
    }

    fun getAllEnemy(): Set<Enemy> {
        val l: MutableSet<Enemy> = TreeSet<Enemy>()
        for (dat in datas) l.addAll(dat!!.enemy.get().possible)
        for (e in getSummon()) l.addAll(e.getPossible())
        return l
    }

    fun getSimple(): Array<common.util.stage.SCDef.Line?> {
        return datas
    }

    fun getSimple(i: Int): common.util.stage.SCDef.Line? {
        return datas[i]
    }

    fun getSMap(): Array<Map.Entry<PackData.Identifier<AbEnemy>, Int>> {
        return smap.entries.toTypedArray<Map.Entry<*, *>>()
    }

    fun getSummon(): Set<AbEnemy> {
        val ans: MutableSet<AbEnemy> = TreeSet<AbEnemy>()
        val temp: MutableSet<AbEnemy> = TreeSet<AbEnemy>()
        val pre: MutableSet<Enemy> = TreeSet<Enemy>()
        val post: MutableSet<Enemy> = TreeSet<Enemy>()
        for (line in datas) {
            val e: AbEnemy? = line!!.enemy.get()
            if (e != null) pre.addAll(e.getPossible())
        }
        while (pre.size > 0) {
            for (e in pre) temp.addAll(e.de.getSummon())
            ans.addAll(temp)
            post.addAll(pre)
            pre.clear()
            for (e in temp) pre.addAll(e.getPossible())
            pre.removeAll(post)
            temp.clear()
        }
        return ans
    }

    fun isSuitable(pack: PackDesc): Boolean {
        for (ints in datas) {
            if (ints!!.enemy.pack == PackData.Identifier.Companion.DEF) continue
            var b = ints.enemy.pack == pack.id
            for (rel in pack.dependency) b = b or (ints.enemy.pack == rel)
            if (!b) return false
        }
        return true
    }

    fun isTrail(): Boolean {
        for (data in datas) if (data!!.castle_0 > 100) return true
        return false
    }

    companion object {
        @Deprecated("")
        val SIZE = 14

        @Deprecated("")
        val E = 0

        @Deprecated("")
        val N = 1

        @Deprecated("")
        val S0 = 2

        @Deprecated("")
        val R0 = 3

        @Deprecated("")
        val R1 = 4

        @Deprecated("")
        val C0 = 5

        @Deprecated("")
        val L0 = 6

        @Deprecated("")
        val L1 = 7

        @Deprecated("")
        val B = 8

        @Deprecated("")
        val M = 9

        @Deprecated("")
        val S1 = 10

        @Deprecated("")
        val C1 = 11

        @Deprecated("")
        val G = 12

        @Deprecated("")
        val M1 = 13
        fun zread(`is`: InStream): SCDef? {
            val t: Int = `is`.nextInt()
            val ver: Int = Data.Companion.getVer(`is`.nextString())
            if (t == 0) {
                if (ver >= 402) {
                    var n: Int = `is`.nextInt()
                    val m: Int = `is`.nextInt()
                    val scd = SCDef(n)
                    val tmp = IntArray(SIZE)
                    for (i in 0 until n) {
                        Arrays.fill(tmp, 0)
                        for (j in 0 until m) tmp[j] = `is`.nextInt()
                        if (m < 14) tmp[M1] = tmp[M]
                        scd.datas[i] = common.util.stage.SCDef.Line(tmp)
                    }
                    scd.sdef = `is`.nextInt()
                    n = `is`.nextInt()
                    for (i in 0 until n) scd.smap[PackData.Identifier.Companion.parseInt<AbEnemy>(`is`.nextInt(), Enemy::class.java)] = `is`.nextInt()
                    n = `is`.nextInt()
                    for (i in 0 until n) {
                        val scg: SCGroup = SCGroup.Companion.zread(`is`)
                        scd.sub.set(scg!!.id, scg)
                    }
                    return scd
                }
            }
            return null
        }
    }
}
